File: 백준/Silver/1182. 부분수열의 합/부분수열의 합.java
import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.InputStreamReader;
import java.io.OutputStreamWriter;

class Main {
    private static int[] nums;
    private static int ans;

    public static void main(String args[]) throws Exception {
        BufferedReader bufferedReader = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bufferedWriter = new BufferedWriter(new OutputStreamWriter(System.out));

        String[] NS = bufferedReader.readLine().split(" ");
        String[] numStr = bufferedReader.readLine().split(" ");
        int N = Integer.parseInt(NS[0]);
        int S = Integer.parseInt(NS[1]);
        nums = new int[N];

        for (int i = 0; i < N; i++) {
            nums[i] = Integer.parseInt(numStr[i]);
        }
        
        back(0, S, 0);

        if (S == 0) ans--;
        
        bufferedWriter.write(String.valueOf(ans));
        bufferedWriter.flush();
    }

    private static void back(int position, int S, int count) {
        if (position == nums.length) {
            if (count == S) {
                ans++;
            }
            
            return;
        }

        back(position + 1, S, count);
        back(position + 1, S, count + nums[position]);
    }
}

File: 백준/Silver/1182. 부분수열의 합/부분수열의 합.py
def back(position, S, count, nums, ans):
    if position == len(nums):
        if count == S:
            ans[0] += 1
        return

    back(position + 1, S, count, nums, ans)  # 원소를 포함하지 않는 경우
    back(position + 1, S, count + nums[position], nums, ans)  # 원소를 포함하는 경우

def main():
    input_data = sys.stdin.read().splitlines()  # 전체 입력을 한 번에 읽음

    N, S = map(int, input_data[0].split())
    nums = list(map(int, input_data[1].split()))

    ans = [0]  # 리스트를 사용하여 참조를 통해 ans를 업데이트

    back(0, S, 0, nums, ans)

    # S가 0인 경우 공집합을 제외하기 위해 ans를 감소시킴
    if S == 0:
        ans[0] -= 1

    print(ans[0])

if __name__ == "__main__":
    import sys
    main()
